lambdas/initialise-ipv-session/src/main/java/uk/gov/di/ipv/core/initialiseipvsession/InitialiseIpvSessionHandler.java

```diff
@@ -321,29 +329,98 @@ private static boolean isListEmpty(List<String> list) {
-                storeHmrcInheritedIdentity(claimsSet, ipvSessionItem, inheritedIdentityVc);
+            if (configService.enabled(EVCS_READ_ENABLED)
+                    && configService.enabled(EVCS_WRITE_ENABLED)) {
+                storeInheritedIdentity(
+                        userId, ipvSessionItem, clientOAuthSessionItem, inheritedIdentityVc);
             }
-        } catch (VerifiableCredentialException | ParseException | UnrecognisedVotException e) {
+        } catch (VerifiableCredentialException | UnrecognisedVotException e) {
             throw new RecoverableJarValidationException(
                     INVALID_INHERITED_IDENTITY_ERROR_OBJECT.setDescription(
                             "Inherited identity JWT failed to validate"),
                     claimsSet,
                     e);
         } catch (JarValidationException e) {
             throw new RecoverableJarValidationException(e.getErrorObject(), claimsSet, e);
+        } catch (EvcsServiceException e) {
+            throw new RecoverableJarValidationException(
+                    INVALID_INHERITED_IDENTITY_ERROR_OBJECT.setDescription(
+                            "Inherited identity VC failed to store"),
+                    claimsSet,
+                    e);
+        } catch (InheritedIdentityException e) {
+            throw new RecoverableJarValidationException(
+                    INVALID_INHERITED_IDENTITY_ERROR_OBJECT.setDescription(
+                            "Too many current inherited identities found in EVCS"),
+                    claimsSet,
+                    e);
+        }
+    }
+
+    @Tracing
+    private void storeInheritedIdentity(
+            String userId,
+            IpvSessionItem ipvSessionItem,
+            ClientOAuthSessionItem clientOAuthSessionItem,
+            VerifiableCredential inheritedIdentityVc)
+            throws EvcsServiceException, CredentialParseException, InheritedIdentityException {
+        var existingInheritedIdentity =
+                evcsService
+                        .getVerifiableCredentials(
+                                userId, clientOAuthSessionItem.getEvcsAccessToken(), CURRENT)
+                        .stream()
+                        .filter(vc -> HMRC_MIGRATION.equals(vc.getCri()))
+                        .toList();
+
+        if (existingInheritedIdentity.size() > 1) {
+            throw new InheritedIdentityException(
+                    String.format(
+                            "Too many current inherited identities found - %d",
+                            existingInheritedIdentity.size()));
+        }
+
+        if (existingInheritedIdentity.isEmpty()) {
+            LOGGER.info("No existing inherited identity found - storing new one");
+
+            evcsService.storeInheritedIdentity(userId, inheritedIdentityVc, null);
+            ipvSessionItem.setInheritedIdentityReceivedThisSession(true);
+            ipvSessionService.updateIpvSession(ipvSessionItem);
```

Comment:
Ooo interesting. It would be good to have a proper understanding of what is an acceptable wait between writing and reading. Happy to park it though until we're closer to inherited identities going live

Reply:
It's not super precise atm, in 'most' cases it'll be very small, but during their load test it occasionally creeps up to a few seconds.

I think the rule of thumb is that it's probably ok if there's some user interaction between the two, but likely to be a problem if it's all internal processing.